src/internal/RangeCounting.cpp

```diff
@@ -0,0 +1,286 @@
+  int depth1 = depth;
+  change = moveToFirstAfter(u, it, depth, 1) || change;
+  insert(l, it1, depth1);
+  --depth;
+  insert(u, it, depth);
+  depth = 0;
+  it = limits_.begin();
+  while(it != limits_.end()) // ideally from it1 to it, but the second insert can invalidate it1
+  {
+    depth -= it->type_;
+    if(depth == 0 && it->type_ == Limit::Cut)
+      it = limits_.erase(it);
+    else
+      ++it;
+  }
+
+  return recompute(change);
+}
+
+const std::vector<Range> & RangeCounting::ranges(bool splitOncountDiff) const
+{
+  if(recompute_ || recomputeSplit_)
+  {
+    intervals_.clear();
+    int depth = 0;
+    if(limits_.empty())
+    {
+      recompute_ = false;
+      recomputeSplit_ = false;
+      return intervals_;
+    }
+    if(splitOncountDiff)
+    {
+      auto next = [&depth, this](auto & it) {
+        const Limit & v = *it;
+        while(it != limits_.end() && *it == v)
+        {
+          depth -= it->type_;
+          ++it;
+        }
+      };
+      auto s = limits_.cbegin();
+      auto it = s;
+      next(it);
+      while(it != limits_.end())
+      {
+        if(s->i_ != it->i_)
+          intervals_.emplace_back(s->i_, it->i_ - s->i_);
+        else
+          assert((s->type_ == Limit::Lower && it->type_ == Limit::Cut)
+                 || (s->type_ == Limit::Cut && it->type_ == Limit::Upper));
+        s = it;
+        next(it);
+        if(depth == 0 && it != limits_.end())
+        {
+          s = it;
+          next(it);
+        }
+      }
+      recompute_ = true;
+      recomputeSplit_ = false;
+    }
+    else
+    {
+      auto s = limits_.cbegin();
+      for(auto it = limits_.cbegin(); it != limits_.cend(); ++it)
+      {
+        depth -= it->type_;
+        if(depth == 0)
+        {
+          intervals_.emplace_back(s->i_, it->i_ - s->i_);
+          s = std::next(it);
+        }
+      }
+      recompute_ = false;
+      recomputeSplit_ = true;
+    }
+  }
+
+  return intervals_;
+}
+
+const std::list<RangeCounting::Limit> & RangeCounting::limits() const { return limits_; }
+
+bool RangeCounting::moveToFirstAfter(const Limit & val, It & it, int & depth, int depthCut) const
+{
+  bool hit = false;
+  while(it != limits_.end() && *it <= val)
+  {
+    depth -= it->type_;
+    if(depth < depthCut)
+      throw std::runtime_error("Too many upper limis. Are you trying to remove values that are not present?");
+    if(depth == depthCut)
+      hit = true;
+    ++it;
+  }
+  return hit;
+}
+
+void RangeCounting::insert(const Limit & val, It & it, int & depth)
+{
+  if(val.type_ == Limit::Lower)
+  {
+    if(val.i_ == it->i_)
+    {
+      assert(it->type_ != Limit::Lower);
+      if(it->type_ == Limit::Cut)
+      {
+        if(auto n = std::next(it); n != limits_.end() && *n == Limit(val.i_, Limit::Upper))
+        {
+          // insertion of (i,+) before (i,|) (i,-): (i,+) is not inserted and (i,-) is removed.
```

Comment:
This comments doesn't explain the special case for `depth == 0`?

Reply:
Fixed in 32599c3